File: data-jpa/src/main/java/top/bettercode/summer/data/jpa/web/PageController.java
package top.bettercode.summer.data.jpa.web;

import java.util.Collection;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.data.web.SpringDataWebProperties;
import org.springframework.data.domain.Page;
import org.springframework.http.ResponseEntity;
import top.bettercode.summer.web.BaseController;
import top.bettercode.summer.web.PagedResources;
import top.bettercode.summer.web.PagedResources.PageMetadata;
import top.bettercode.summer.web.RespExtra;

/**
 * @author Peter Wu
 */
public class PageController extends BaseController {

  @Autowired
  private SpringDataWebProperties properties;

  @Override
  protected RespExtra<?> of(Object object) {
    return super.of(pagedObject(object));
  }

  @Override
  protected ResponseEntity<?> ok(Object object) {
    if (object instanceof Page) {
      return super.ok(pagedResources((Page<?>) object));
    } else {
      return super.ok(object);
    }
  }

  protected <T, R> ResponseEntity<?> ok(Page<T> object, Function<? super T, ? extends R> mapper) {
    int number =
        properties.getPageable().isOneIndexedParameters() ? object.getNumber() + 1
            : object.getNumber();
    List<T> content = object.getContent();
    return super.ok(
        new PagedResources<>(new PageMetadata(number, object.getSize(), object.getTotalPages(),
            object.getTotalElements()),
            content.stream().map(mapper).collect(Collectors.toList())));
  }

  protected ResponseEntity<?> page(Object object) {
    return super.ok(pagedObject(object));
  }

  private Object pagedObject(Object object) {
    if (object instanceof Page) {
      return pagedResources((Page<?>) object);
    } else if (object instanceof Collection) {
      Collection<?> collection = (Collection<?>) object;
      int number = properties.getPageable().isOneIndexedParameters() ? 1 : 0;
      int size = collection.size();
      return new PagedResources<>(new PageMetadata(number, size, 1, size), collection);
    } else if (object != null && object.getClass().isArray()) {
      Object[] array = (Object[]) object;
      int number = properties.getPageable().isOneIndexedParameters() ? 1 : 0;
      int size = array.length;
      return new PagedResources<>(new PageMetadata(number, size, 1, size), array);
    } else {
      return object;
    }
  }

  private PagedResources<?> pagedResources(Page<?> object) {
    int number =
        properties.getPageable().isOneIndexedParameters() ? object.getNumber() + 1
            : object.getNumber();
    return new PagedResources<>(new PageMetadata(number, object.getSize(), object.getTotalPages(),
        object.getTotalElements()), object.getContent());
  }

}

File: data-jpa/src/main/kotlin/top/bettercode/summer/data/jpa/web/PageController.kt
package top.bettercode.summer.data.jpa.web

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.data.web.SpringDataWebProperties
import org.springframework.data.domain.Page
import org.springframework.http.ResponseEntity
import top.bettercode.summer.web.BaseController
import top.bettercode.summer.web.PagedResources
import top.bettercode.summer.web.RespExtra
import java.util.function.Function
import java.util.stream.Collectors

/**
 * @author Peter Wu
 */
open class PageController : BaseController() {

    @Autowired
    private lateinit var properties: SpringDataWebProperties

    @JvmOverloads
    protected fun <T, R> ok(
        `object`: Page<T?>,
        mapper: Function<T?, R?>? = null
    ): ResponseEntity<*> {
        return super.ok(pagedResources(`object`, mapper))
    }

    protected fun <T, R> ok(`object`: Collection<T?>, mapper: Function<T?, R?>): ResponseEntity<*> {
        return super.ok(`object`.stream().map(mapper).collect(Collectors.toList()))
    }

    protected fun <T, R> ok(`object`: Array<T?>, mapper: Function<T?, R?>): ResponseEntity<*> {
        return super.ok(`object`.toList().stream().map(mapper).collect(Collectors.toList()))
    }

    protected fun page(`object`: Any?): ResponseEntity<*> {
        return super.ok(`object`)
    }

    @JvmOverloads
    protected fun <T, R> page(
        `object`: Page<T?>,
        mapper: Function<T?, R?>? = null
    ): ResponseEntity<*> {
        return super.ok(pagedResources(`object`, mapper))
    }

    @JvmOverloads
    protected fun <T, R> page(
        `object`: Collection<T?>,
        mapper: Function<T?, R?>? = null
    ): ResponseEntity<*> {
        val number = if (properties.pageable.isOneIndexedParameters) 1 else 0
        val size = `object`.size
        val collect = if (mapper == null) `object` else `object`.stream().map(mapper)
            .collect(Collectors.toList())
        val pagedResources =
            pagedResources(number, size, 1, size.toLong(), collect)
        return super.ok(pagedResources)
    }

    @JvmOverloads
    protected fun <T, R> page(
        `object`: Array<T?>,
        mapper: Function<T?, R?>? = null
    ): ResponseEntity<*> {
        val number = if (properties.pageable.isOneIndexedParameters) 1 else 0
        val size = `object`.size
        val collect =
            if (mapper == null) `object`.toList() else `object`.toList().stream().map(mapper)
                .collect(Collectors.toList())
        val pagedResources =
            pagedResources(number, size, 1, size.toLong(), collect)
        return super.ok(pagedResources)
    }

    @JvmOverloads
    fun <T, R> of(`object`: Page<T?>, mapper: Function<T?, R?>? = null): RespExtra<*>? {
        return super.of(pagedResources(`object`, mapper))
    }

    @JvmOverloads
    fun <T, R> of(`object`: Collection<T?>, mapper: Function<T?, R?>? = null): RespExtra<*>? {
        val number = if (properties.pageable.isOneIndexedParameters) 1 else 0
        val size = `object`.size
        val collect = if (mapper == null) `object` else `object`.stream().map(mapper)
            .collect(Collectors.toList())
        val pagedResources =
            pagedResources(number, size, 1, size.toLong(), collect)
        return super.of(pagedResources)
    }

    @JvmOverloads
    fun <T, R> of(`object`: Array<T?>, mapper: Function<T?, R?>? = null): RespExtra<*>? {
        val number = if (properties.pageable.isOneIndexedParameters) 1 else 0
        val size = `object`.size
        val collect =
            if (mapper == null) `object`.toList() else `object`.toList().stream().map(mapper)
                .collect(Collectors.toList())
        val pagedResources =
            pagedResources(number, size, 1, size.toLong(), collect)
        return super.of(pagedResources)
    }

    @JvmOverloads
    protected fun <T, R> pagedResources(`object`: Page<T?>, mapper: Function<T?, R?>? = null): Any {
        val number =
            if (properties.pageable.isOneIndexedParameters) `object`.number + 1 else `object`.number
        val content =
            if (mapper == null) `object`.content else `object`.content.stream().map(mapper)
                .collect(Collectors.toList())
        return pagedResources(
            number,
            `object`.size,
            `object`.totalPages,
            `object`.totalElements,
            content
        )
    }

    protected open fun <T> pagedResources(
        number: Int,
        size: Int,
        totalPages: Int,
        totalElements: Long,
        content: Collection<T?>
    ): Any {
        return PagedResources(number, size, totalPages, totalElements, content)
    }
}
